### ISSUE TITLE
As a CIF user, I want to add contacts in the system, so that I can add them to a project

### ISSUE THREAD

**Issue by matthieu-foucault** [bcgov/cas-cif#159]
#### Description:

A contact may be linked to more than one project. Contact updates should not result in a funding agreement amendment, so a contact's info can be updated for all of its projects at the same time.

To allow reusing contacts, the user should be able to manage a single list of contacts for the whole application.

Although consultants can be the ones filling a CIF proposal, the contacts are employees of the operator at the reporting operation, so a contact can be associated to an operator. We can leave this association optional for greater flexibility.

#### Acceptance Criteria:

Given that I am logged in as a CIF user
When I go to the `cif/contacts` page
Then I can see a list of the existing contacts

Given that I am logged in as a CIF user
When I click on a contact list item in the contacts list view
Then I can see a contact detail view showing more detailed contact information

Given that I am logged in as a CIF user and am looking at a contact detail view
When I click on the edit contact button
Then I can see a contact edit form that lets me update and save contact information

#### Development Checklist:
- [x] the `contact` table exists
- [x] the `contact` table has a `given_name` column
- [x] the `contact` table has a `family_name` column
- [x] the `contact` table has a `email_address` column
- [x] the `contact` table has a `phone_number` column
- [x] the `contact` table has a `phone_number_ext` column
- [x] the `contact` table has a `position` column
- [x] the `contact` table has a `comments` column
- [x] the `contact_project` table has two `project_id` and `contact_id` foreign keys
- [x] Create a contact detail view that is shown when a contact row item is clicked
- [x] Create a contact edit view that allows editing and saving of contact information
  - operators and project cannot be edited from the contact page
~- [ ] Show a list of all associated projects to that contact on the contact detail view~ (may add later if needed)

#### Notes:
-
#### Personas:
- [Program Development Team](https://bcgov.sharepoint.com/:b:/t/00608-CIFInternalUsers/EcVHGc4X8mhIvQHGSaGJmyAB9F1iBe7MSH1Y1fBKUc1skA?e=fzATIc)
- [Operations Team](https://bcgov.sharepoint.com/:b:/t/00608-CIFInternalUsers/EaswKcprmuNKpEuY29a-Os0BG-h5u-Ztou7UuzK3x0xfjw?e=e8oR1i)

**Comment by rajpersram** [comment]
![CIF Workflow Co-Design - Edit Project - Contacts.jpg](https://images.zenhubusercontent.com/5c2e4cfdab4f3464fb8df8d1/630d5d66-c57f-4cf9-9bcd-09c98bc5e9bb)

**Comment by rajpersram** [comment]
@Peirist Taylor, can you confirm whether or not CIF needs to be able to associate a given contact with more than one Operator? Is this an edge case, or typical? Thanks.

**Comment by Peirist** [comment]
This is not typical but it is possible; particularly in cases where proposals are being prepared by third-party consultants who may end up doing work for multiple reporting operations in the same intake.  It would be ideal if the system is built to accommodate this possibility; the alternative would be having multiple entries of the same contact assigned to different operators.

**Comment by matthieu-foucault** [comment]
@Peirist Follow up question, which would affect @rajpersram 's wireframes. Would you rather:

1. Create contacts, assign them to projects, and then have a way to see contacts per operator (by going through the contact-project and project-operator associations), or,

1. Create contacts, assign them to operators, and then, when creating a project, you can add contacts to the project, providing that they are in the project operator's contacts list

In other words, is there a scenario where you would want a contact in the system that is associated with an operator, but there is no CIF project for that operator and contact?

**Comment by Peirist** [comment]
For where our scope of work is now, I think option 1 is preferable; we should focus on projects and their associated contacts rather than the other way around.  I will note that having a broader (and maybe entirely separate) list of contacts at some point may be an objective - we have contacts associated with each operator for the purposes of updating on the program generally, and these contacts are typically not the contacts assigned to projects (which tend to be engineers, project managers, etc.).

**Comment by matthieu-foucault** [comment]
@pbastia  are you ok with 5pts or do you think that this should be 8?

**Comment by LindsayMacfarlane** [comment]
Thank you @matthieu-foucault! I tested in dev:

- Adding a contact from the contacts page: I couldn't type in any of the fields. When I refreshed the page, the order of the fields also changed. Perhaps I tested to soon?

**Comment by matthieu-foucault** [comment]
There's definitely a bug in there. Back to "in progress"

**Comment by LindsayMacfarlane** [comment]
Tested in dev:
- searching for a contact in the contact table. Table filters as expected. If no results, user get appropriate "no results found" message. Clear button clears filter as expected. 
- Able to enter a contact
- Able to view contact information. Note: on the view page, comments are not currently being displayed. 
- Was able to edit a contact. Note: in my edit, I added the contact position. This updated in the view, but did not update in the contact search table when I first went to the search page. However, it did display when I went back later.  
- Was able to select the contact I created when creating a new project.

**Comment by matthieu-foucault** [comment]
> This updated in the view, but did not update in the contact search table when I first went to the search page. However, it did display when I went back later.

This will be fixed with #285, which is in our current sprint's backlog

**Comment by LindsayMacfarlane** [comment]
Awesome! Thank you @matthieu-foucault !
